feat(breadcrumbs): migrate to TS & reorg file structure

hzhu opened the pull request:
## Description

This pull request migrates the breadcrumbs package to TypeScript, reorganizes the folder structure, and updates the breadcrumb React Styleguidist example.

## Details

I added a Basic example under Examples in React Styleguidist and did not include advanced examples or knobs. If there are ideas for advanced examples or knobs, I would like to see how helpful they are for consumers and get them added.

A review link for the breadcrumb package Storybook is published [**here**](https://5dd5cc4d717a0867cb61cd15--gallant-blackwell-63466f.netlify.com/breadcrumbs) (updated).

The new file organization is:

```
/examples
  basic.md
/src
  /elements
    Breadcrumb.tsx
  /styled
    StyledBreadcrumb.ts
    StyledBreadcrumbItem.ts
```

## Checklist

- [x] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer). [Review link](https://5dd5cc4d717a0867cb61cd15--gallant-blackwell-63466f.netlify.com/breadcrumbs).
- [x] :nail_care: view component styling is based on a Garden CSS
      component
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -37,18 +40,29 @@ const renderItems = (items, getCurrentPageProps) => {
 };
 
+export interface IBreadCrumbProps extends React.HTMLProps<HTMLDivElement> {
+  children?: React.ReactNode;
```
The `children` prop should be created for us by the `React.FC` call. This interface may not be needed.

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -37,18 +40,29 @@ const renderItems = (items, getCurrentPageProps) => {
+
 /** @component */
-export default function Breadcrumb({ children, ...breadcrumbProps }) {
+const Breadcrumb: React.FunctionComponent<IBreadCrumbProps> = ({
```
Since this element is publicly consumable, this should be using `React.forwardRef`

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -37,18 +40,29 @@ const renderItems = (items, getCurrentPageProps) => {
-      <List {...breadcrumbProps}>{renderItems(children, getCurrentPageProps)}</List>
-    </BreadcrumbView>
+    /* role not needed as `StyledBreadcrumb` is a navigation landmark. */
+    <StyledBreadcrumb {...getContainerProps({ role: null } as any)}>
```
I believe we removed the need to apply `{role: null}` in your recent a11y sweep of react-containers. If this is a `nav` element it may no longer be needed.

hzhu replied:
I believe this is still needed. The recent [sweep](https://github.com/zendeskgarden/react-containers/pull/129) 🧹was regarding `aria-orientation` for `react-selection`.

We may need to do an additional sweep. Maybe we can sync on this after this PR.

austin94 commented on packages/breadcrumbs/src/elements/Item.ts:
```diff
@@ -13,10 +13,14 @@ import BreadcrumbStyles from '@zendeskgarden/css-breadcrumbs';
 const COMPONENT_ID = 'breadcrumbs.item';
 
+export interface IItemProps extends React.HTMLProps<HTMLLIElement> {
+  current?: boolean;
```
Boolean props are supposed to use the `is*` naming now. Change to `isCurrent`

austin94 commented on packages/breadcrumbs/src/elements/Item.ts:
```diff
@@ -13,10 +13,14 @@ import BreadcrumbStyles from '@zendeskgarden/css-breadcrumbs';
  * Accepts all `<li>` props
  */
-const Item = styled.li.attrs(props => ({
+const Item = styled.li.attrs((props: IItemProps) => ({
```
Styled-components can handle the generics for you here. Try

```tsx
const Item = styled.li.attrs<IItemProps>(...)<IItemProps>``;
```

austin94 commented on packages/breadcrumbs/src/elements/Item.ts:
```diff
@@ -25,11 +29,13 @@ const Item = styled.li.attrs(props => ({
 }))`
   ${props => retrieveComponentStyles(COMPONENT_ID, props)};
-`;
+` as React.FC<IItemProps>;
```
This cast shouldn't be necessary (and removes the ability to use `ref`)

austin94 commented on packages/breadcrumbs/src/elements/Item.ts:
```diff
@@ -13,10 +13,14 @@ import BreadcrumbStyles from '@zendeskgarden/css-breadcrumbs';
+const Item = styled.li.attrs((props: IItemProps) => ({
   'data-garden-id': COMPONENT_ID,
   'data-garden-version': PACKAGE_VERSION,
   className: classNames(BreadcrumbStyles['c-breadcrumb__item'], {
```
All styled-components should be within the `styled` folder

austin94 commented on packages/breadcrumbs/src/elements/List.ts:
```diff
@@ -12,6 +12,10 @@ import { retrieveComponentStyles, isRtl } from '@zendeskgarden/react-theming';
 
 const COMPONENT_ID = 'breadcrumbs.list';
 
+export interface IListProps extends React.HTMLProps<HTMLDivElement> {
```
This interface isn't needed.

austin94 commented on packages/breadcrumbs/src/elements/List.ts:
```diff
@@ -24,7 +28,7 @@ const List = styled.ol.attrs(props => ({
 }))`
   ${props => retrieveComponentStyles(COMPONENT_ID, props)};
-`;
+` as React.FC<IListProps>;
```
We shouldn't down-cast styled-components. This removes the ability to use `ref`

austin94 commented on packages/breadcrumbs/src/styled/StyledBreadcrumb.ts:
```diff
@@ -13,12 +13,12 @@ const COMPONENT_ID = 'breadcrumbs.breadcrumb_view';
  * Accepts all `<nav>` props
  */
-const BreadcrumbView = styled.nav.attrs({
+const StyledBreadcrumb = styled.nav.attrs({
```
For all styled-components we are using named exports. Change this to `export const StyledBreadcrumb`

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -38,17 +41,17 @@ const renderItems = (items, getCurrentPageProps) => {
 
 /** @component */
-export default function Breadcrumb({ children, ...breadcrumbProps }) {
+const Breadcrumb: React.FC = ({ children, ...breadcrumbProps }) => {
```
This FC can also have an additional type interface.

```
React.FC<HTMLAttributes<HTMLDivElement>> = (...)
```

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -0,0 +1,34 @@
+import React, { Children, cloneElement, HTMLAttributes } from 'react';
+import { useBreadcrumb } from '@zendeskgarden/container-breadcrumb';
+
+import { StyledBreadcrumb } from '../styled/StyledBreadcrumb';
```
Could we import these from the single `styled/index.ts` file? Something like

```tsx
import { StyledBreadcrumb, StyledItem } from '../styled';
```

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -0,0 +1,34 @@
+import { StyledBreadcrumb } from '../styled/StyledBreadcrumb';
+import { StyledItem } from '../styled/StyledItem';
+
+export const Breadcrumb: React.FC<HTMLAttributes<HTMLOListElement>> = ({ children, ...props }) => {
```
This should be using `React.forwardRef` to supply the `<nav>` ref to any consumers.

austin94 commented on packages/breadcrumbs/src/index.js:
```diff
@@ -6,6 +6,6 @@
-export { default as BreadcrumbView } from './views/BreadcrumbView';
-export { default as List } from './views/List';
-export { default as Item } from './views/Item';
+export { default as BreadcrumbView } from './styled/BreadcrumbView';
```
We no longer export any of the `styled` elements to help reduce API churn. Let's remove `BreadcrumbView`, `List`, and `Item.

austin94 replied:
Also, looks like this file should still be moved to `TS`. This would correct some naming errors here as well.

hzhu replied:
Oh these comments were blazing fast! I saw this during my second review and addressed with new commit. I'll go ahead and address other comments too. I'll re-request a review once resolve them 👍 .

austin94 commented on packages/breadcrumbs/src/styled/StyledBreadcrumb.spec.tsx:
```diff
@@ -7,17 +7,17 @@
+import { StyledBreadcrumb } from './StyledBreadcrumb';
 
-describe('List', () => {
+describe('StyledList', () => {
```
Should be `StyledBreadcrumb`

austin94 commented on packages/breadcrumbs/src/styled/StyledItem.ts:
```diff
@@ -5,31 +5,24 @@
+  isCurrent?: boolean;
+}
+
+export const StyledItem = styled.li.attrs<IStyledItemProps>(props => ({
```
Could we name this to `StyledBreadcrumbItem` to be more specific?

hzhu commented on packages/breadcrumbs/examples/basic.md:
```diff
@@ -1,6 +1,6 @@
 The `Breadcrumb` component follows the
```
👀

hzhu replied:
To be consistent with other packages:

> The Breadcrumb component acts as a high-abstraction API over the [`useBreadcrumb()`](https://garden.zendesk.com/react-components/breadcrumbs/) hook.

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -0,0 +1,39 @@
+
+import { StyledBreadcrumb, StyledBreadcrumbItem } from '../styled';
+
+export const Breadcrumb: React.FC<HTMLAttributes<HTMLOListElement>> = React.forwardRef(
```
This is downcasting the `forwardRef` value and removes the ability for the `ref` to be used. `forwardRef` provides some generics for us to use. Try:

```tsx
export const Breadcrumb = React.fowardRef<HTMLElement, HTMLAttributes<HTMLElement>((props, ref) => {});
```

This will type the props and ref value correctly.

austin94 replied:
Also, now that the props are spread onto a `<nav>` element I believe `HTMLElement` is the correct base-type, not `HTMLOListElement`

hzhu replied:
Hmmm, thanks I'll give that a shot!

hzhu replied:
Thanks for the suggestion! This makes sense.

austin94 commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -0,0 +1,39 @@
+    });
+
+    return (
+      <nav ref={ref} {...getContainerProps({ ...props, role: null } as any)}>
```
Lets pass `ref` through `getContainerProps()` as well for consistency

austin94 commented on packages/breadcrumbs/src/index.ts:
```diff
@@ -6,6 +6,3 @@
  */
 
 export { default as Breadcrumb } from './elements/Breadcrumb';
```
I think this should be a named export, not default.

jzempel commented:
oh, @hzhu plz remember to fill out the `BREAKING_CHANGES.md` since `Item` is being removed, etc.

jzempel commented on BREAKING_CHANGES.md:
```diff
@@ -4,6 +4,10 @@ NOTE: many style-related COMPONENT_IDs were renamed for consistency based
 
+## @zendeskgarden/react-breadcrumbs
+
+- `<Item>` has been removed.
```
A little explainer for how the updated breadcrumb handles children would go a long way.

hzhu replied:
Oh yeah, that would be more helpful. How does this sound?

> The `<Item>` component has been removed. The `<Breadcrumb>` component determines the last breadcrumb item and applies styling to each item. Instead of passing `<Item>` components as children to `<Breadcrumb>`, consumers can now pass in inline elements of their choice. For example, a child of `<Breadcrumb>` can be a `<span>` or a component that renders an `<a>`.

jzempel replied:
brilliant.

hzhu commented on packages/breadcrumbs/src/elements/Breadcrumb.tsx:
```diff
@@ -0,0 +1,39 @@
+
+import { StyledBreadcrumb, StyledBreadcrumbItem } from '../styled';
+
+export const Breadcrumb = React.forwardRef<HTMLElement, HTMLAttributes<HTMLElement>>(
```
Since this component now forwards a ref, there should be a test for it, [similar to other packages](https://github.com/zendeskgarden/react-components/pull/529/files#diff-a658bb64292d200e01d13327f68d34d1R20).

hzhu replied:
Test added 👍.

hzhu commented:
@zendeskgarden/maintainers there may be a new chevron icon that will be used in the Breadcrumbs component. If so, that will be reflected immediately in the next PR for the breadcrumbs package: CSS-in-JS 👍rather than this TypeScript migration / file re-org PR. 

(Thanks @ginnywood for reminding me :)

hzhu commented:
> It looks like there is a linting error in a `spec` file and a rebase conflict in `BREAKING_CHANGES`.
> 
> Did the linting error slip through the `pre-commit` hooks? We might need to update those if so

Yesterday, the `pre-commit` hook yielded a lint error for two lines on `BREAKING_CHANGES`. The character length for the lines exceeded 100 characters, so I adjusted them accordingly to fit 100 characters per line.

I rebased my changes on top of `next` — the build should be passing now ✅.
